services/src/panic.c

```diff
@@ -46,29 +55,48 @@ void panic_set_hook(const PanicHook panicHookFunction, void* reserved)
     panicHook = panicHookFunction;
 }
 
-void panic_(const ePanicCode code, void* extraInfo, void(*unnamed)(uint32_t))
+void panic_(const ePanicCode code, const char* text, void* unused) {
+    PANIC_COMPAT(code, text);
+}
+
+void panic_ext(const PanicData* data, void* reserved)
 {
     #if HAL_PLATFORM_CORE_ENTER_PANIC_MODE
         HAL_Core_Enter_Panic_Mode(NULL);
     #else
         HAL_disable_irq();
     #endif // HAL_PLATFORM_CORE_ENTER_PANIC_MODE
 
+    memset(&g_retainedPanicData, 0, sizeof(g_retainedPanicData));
+    g_retainedPanicData.marker = PANIC_DATA_MARKER;
+    memcpy(&g_retainedPanicData.data, data, sizeof(PanicData));
+
     //run the panic hook (if present)! this func doens't have to
     //return and can handle the system state / exit method on its own if required
     if( panicHook ) {
-        panicHook(code, extraInfo);
+        panicHook(data->code, data->text);
     }
 
     //always run the internal function if the user func returns
     //this flashes the LEDs in the fixed panic pattern
-    panic_internal(code, extraInfo);
+    panic_internal(data->code, data->text);
 
     #if defined(RELEASE_BUILD) || PANIC_BUT_KEEP_CALM == 1
-        HAL_Core_System_Reset_Ex(RESET_REASON_PANIC, code, NULL);
+        HAL_Core_System_Reset_Ex(RESET_REASON_PANIC, data->code, NULL);
     #endif
 }
 
+int panic_get_last_panic_data(PanicData* panic, void* reserved) {
+    CHECK_TRUE(panic, SYSTEM_ERROR_INVALID_ARGUMENT);
+    CHECK_TRUE(g_retainedPanicData.marker == PANIC_DATA_MARKER && g_retainedPanicData.data.size > 0, SYSTEM_ERROR_NOT_FOUND);
+    memcpy(panic, &g_retainedPanicData.data, panic->size <= g_retainedPanicData.data.size ? panic->size : g_retainedPanicData.data.size);
```

Comment:
`panic->size <= g_retainedPanicData.data.size ? panic->size : g_retainedPanicData.data.size`
Is the size check here to account for a growing `PanicData` structure? (not sure how that would work with a panic between firmware updates that change the structure)
Or just a sanity check?